File: ch03/src/cpp/acquire_release.cpp
#include <atomic>
#include <chrono>
#include <cstdint>
#include <iostream>
#include <thread>

namespace {
auto sData = std::uint64_t { 0U };
auto sReady = std::atomic_bool { false };
}

int main()
{
    auto t = std::thread { []() {
        // Modifying sData is safe here because we haven't set the sReady flag
        // => Nothing else is accessing data
        sData = 123;
        sReady.store(true, std::memory_order_release); // Everything before this stores ..
    } };

    while (!sReady.load(std::memory_order_acquire)) { // .. is visible after this loads `true`
        std::this_thread::sleep_for(std::chrono::milliseconds { 100 });
        std::cout << "Waiting ...\n";
    }

    // Accessing sData here is safe because READY is already set,
    // meaning that nothing is mutating DATA
    std::cout << sData << std::endl;

    t.join();

    return 0;
}

File: ch03/src/bin/acquire_release.rs
use std::{
    sync::atomic::{AtomicBool, AtomicU64, Ordering},
    thread,
    time::Duration,
};

static DATA: AtomicU64 = AtomicU64::new(0);
static READY: AtomicBool = AtomicBool::new(false);

fn main() {
    thread::spawn(|| {
        DATA.store(123, Ordering::Relaxed);
        READY.store(true, Ordering::Release); // Everything before this store ...
    });

    while !READY.load(Ordering::Acquire) {
        // ... is visible after this loads `true`
        thread::sleep(Duration::from_millis(100));
        println!("Waiting...");
    }

    println!("{}", DATA.load(Ordering::Relaxed));
}
